File: macos/Sources/BambuDryCore/Control/Orchestrator.swift
import Foundation

/// Connects a LANTransport to one AutoDryController per AMS unit on a printer.
///
/// On each report:
///   1. Decode the AMS array
///   2. For each AMS, derive an AMSSnapshot (auto-detecting model from `info`)
///   3. Throttle per-AMS evaluation to once per `tickInterval`
///   4. Call AutoDryController.evaluate()
///   5. Publish resulting DryRequest (or just log it, in dry-run mode)
///
/// One-orchestrator-per-printer; multiple printers run independent instances.
public actor Orchestrator {

    public struct Config: Sendable {
        public var settingsByAmsId: [Int: AutoDrySettings]   // applies to AMS by id (0, 128, 129…)
        public var defaultSettings: AutoDrySettings           // for AMS not in the map
        public var dryRun: Bool
        public var tickInterval: TimeInterval
        public var pushAllInterval: TimeInterval               // re-request full status periodically
        public init(settingsByAmsId: [Int: AutoDrySettings] = [:],
                    defaultSettings: AutoDrySettings = .init(),
                    dryRun: Bool = true,
                    tickInterval: TimeInterval = 5,
                    pushAllInterval: TimeInterval = 60) {
            self.settingsByAmsId = settingsByAmsId
            self.defaultSettings = defaultSettings
            self.dryRun = dryRun
            self.tickInterval = tickInterval
            self.pushAllInterval = pushAllInterval
        }
    }

    public struct Event: Sendable {
        public enum Kind: Sendable {
            case observed(AMSSnapshot)
            case decision(amsId: Int, AutoDryController.Decision)
            case sent(amsId: Int, json: String)
            case dryRunSkipped(amsId: Int, AutoDryController.Decision)
            case error(String)
        }
        public let timestamp: Date
        public let kind: Kind
    }

    private let transport: LANTransport
    private var config: Config

    private var controllers: [Int: AutoDryController] = [:]
    private var lastEvalAt: [Int: Date] = [:]
    private var lastPushAllAt: Date = .distantPast

    private var eventContinuation: AsyncStream<Event>.Continuation?

    public init(transport: LANTransport, config: Config) {
        self.transport = transport
        self.config = config
    }

    // MARK: - Live config mutation
    //
    // The orchestrator is long-lived; the user can change settings from the
    // menu bar UI after we're already running. These setters let the host
    // app push updates without rebuilding the whole pipeline.

    public func setSettings(_ settings: AutoDrySettings, forAmsId id: Int) {
        config.settingsByAmsId[id] = settings
    }

    public func setDefaultSettings(_ settings: AutoDrySettings) {
        config.defaultSettings = settings
    }

    public func setDryRun(_ dryRun: Bool) {
        config.dryRun = dryRun
    }

    public func events() -> AsyncStream<Event> {
        AsyncStream { continuation in
            self.eventContinuation = continuation
            continuation.onTermination = { [weak self] _ in
                Task { await self?.clearEventContinuation() }
            }
        }
    }

    private func clearEventContinuation() {
        eventContinuation = nil
    }

    private func emit(_ kind: Event.Kind) {
        eventContinuation?.yield(Event(timestamp: Date(), kind: kind))
    }

    /// Drive the loop: connect, request pushall, then process every report.
    public func run() async throws {
        try await transport.connect()
        try transport.requestPushAll()
        lastPushAllAt = Date()

        for await envelope in transport.reportStream() {
            await ingest(envelope)
            await maybeRefresh()
        }
    }

    private func maybeRefresh() async {
        let now = Date()
        if now.timeIntervalSince(lastPushAllAt) >= config.pushAllInterval {
            do { try transport.requestPushAll() } catch {
                emit(.error("pushAll failed: \(error)"))
            }
            lastPushAllAt = now
        }
    }

    private func ingest(_ envelope: ReportEnvelope) async {
        guard let amsList = envelope.print?.ams?.ams else { return }
        let isPrinting = (envelope.print?.gcode_state ?? "") == "RUNNING"
            || (envelope.print?.gcode_state ?? "") == "PRINTING"
            || (envelope.print?.gcode_state ?? "") == "PAUSE"

        for ams in amsList {
            guard let snapshot = ams.snapshot() else { continue }
            emit(.observed(snapshot))

            // Throttle per-AMS evaluation
            let now = Date()
            if let last = lastEvalAt[snapshot.amsId],
               now.timeIntervalSince(last) < config.tickInterval { continue }
            lastEvalAt[snapshot.amsId] = now

            // One controller per AMS, lazily created
            let controller = controllers[snapshot.amsId] ?? AutoDryController(amsId: snapshot.amsId)
            controllers[snapshot.amsId] = controller

            let settings = config.settingsByAmsId[snapshot.amsId] ?? config.defaultSettings
            let decision = await controller.evaluate(snapshot: snapshot,
                                                     settings: settings,
                                                     printerIsPrinting: isPrinting)
            emit(.decision(amsId: snapshot.amsId, decision))

            switch decision {
            case .start(let amsId, let filament, let tempC, let hours):
                if config.dryRun {
                    emit(.dryRunSkipped(amsId: amsId, decision))
                } else {
                    do {
                        let json = try transport.startDrying(amsId: amsId,
                                                             filament: filament,
                                                             tempC: tempC,
                                                             hours: hours)
                        emit(.sent(amsId: amsId, json: json))
                    } catch {
                        emit(.error("start publish failed: \(error)"))
                    }
                }
            case .stop(let amsId):
                if config.dryRun {
                    emit(.dryRunSkipped(amsId: amsId, decision))
                } else {
                    do {
                        let json = try transport.stopDrying(amsId: amsId)
                        emit(.sent(amsId: amsId, json: json))
                    } catch {
                        emit(.error("stop publish failed: \(error)"))
                    }
                }
            case .noop:
                break
            }
        }
    }
}

File: windows/src/BambuDry.Core/Control/Orchestrator.cs
using System.Threading.Channels;
using BambuDry.Core.Net;

namespace BambuDry.Core.Control;

/// <summary>
/// Connects a LanTransport to one AutoDryController per AMS unit on a printer.
///
/// On each report:
///   1. Decode the AMS array
///   2. For each AMS, derive an AmsSnapshot (auto-detecting model from `info`)
///   3. Throttle per-AMS evaluation to once per <see cref="OrchestratorConfig.TickInterval"/>
///   4. Call AutoDryController.Evaluate()
///   5. Publish the resulting DryRequest (or just log it, in dry-run mode)
///
/// One-orchestrator-per-printer; multiple printers run independent instances.
/// </summary>
public sealed class Orchestrator
{
    public abstract record Event(DateTime Timestamp)
    {
        public sealed record Observed(DateTime Timestamp, AmsSnapshot Snapshot) : Event(Timestamp);
        public sealed record DecisionMade(DateTime Timestamp, int AmsId, Decision Decision) : Event(Timestamp);
        public sealed record Sent(DateTime Timestamp, int AmsId, string Json) : Event(Timestamp);
        public sealed record DryRunSkipped(DateTime Timestamp, int AmsId, Decision Decision) : Event(Timestamp);
        public sealed record ErrorOccurred(DateTime Timestamp, string Message) : Event(Timestamp);
    }

    private readonly LanTransport _transport;
    private readonly object _lock = new();
    private OrchestratorConfig _config;

    private readonly Dictionary<int, AutoDryController> _controllers = new();
    private readonly Dictionary<int, DateTime> _lastEvalAt = new();
    private DateTime _lastPushAllAt = DateTime.MinValue;

    private readonly Channel<Event> _events =
        Channel.CreateUnbounded<Event>(new UnboundedChannelOptions { SingleWriter = true });

    public Orchestrator(LanTransport transport, OrchestratorConfig config)
    {
        _transport = transport;
        _config = config;
    }

    public IAsyncEnumerable<Event> Events(CancellationToken ct = default) => _events.Reader.ReadAllAsync(ct);

    // Live config mutation — the orchestrator is long-lived; the user can change
    // settings from the menu bar UI after we're already running.
    public void SetSettings(AutoDrySettings settings, int amsId)
    {
        lock (_lock) _config = _config with { SettingsByAmsId = With(_config.SettingsByAmsId, amsId, settings) };
    }
    public void SetDefaultSettings(AutoDrySettings settings)
    {
        lock (_lock) _config = _config with { DefaultSettings = settings };
    }
    public void SetDryRun(bool dryRun)
    {
        lock (_lock) _config = _config with { DryRun = dryRun };
    }

    private OrchestratorConfig CurrentConfig() { lock (_lock) return _config; }

    /// <summary>Drive the loop: connect, request pushall, then process every report.</summary>
    public async Task RunAsync(CancellationToken ct = default)
    {
        try
        {
            await _transport.ConnectAsync(ct: ct).ConfigureAwait(false);
            await _transport.RequestPushAllAsync(ct).ConfigureAwait(false);
            _lastPushAllAt = DateTime.UtcNow;

            await foreach (var envelope in _transport.ReportStream(ct).ConfigureAwait(false))
            {
                await IngestAsync(envelope, ct).ConfigureAwait(false);
                await MaybeRefreshAsync(ct).ConfigureAwait(false);
            }
        }
        finally
        {
            _events.Writer.TryComplete();
        }
    }

    private async Task MaybeRefreshAsync(CancellationToken ct)
    {
        var now = DateTime.UtcNow;
        var cfg = CurrentConfig();
        if (now - _lastPushAllAt >= cfg.PushAllInterval)
        {
            try { await _transport.RequestPushAllAsync(ct).ConfigureAwait(false); }
            catch (Exception ex) { Emit(new Event.ErrorOccurred(DateTime.UtcNow, $"pushAll failed: {ex.Message}")); }
            _lastPushAllAt = now;
        }
    }

    private async Task IngestAsync(ReportEnvelope envelope, CancellationToken ct)
    {
        var amsList = envelope.Print?.Ams?.Ams;
        if (amsList is null) return;

        var gcodeState = envelope.Print?.GcodeState ?? "";
        var isPrinting = gcodeState is "RUNNING" or "PRINTING" or "PAUSE";

        var cfg = CurrentConfig();

        foreach (var ams in amsList)
        {
            var snapshot = ams.Snapshot();
            if (snapshot is null) continue;

            Emit(new Event.Observed(DateTime.UtcNow, snapshot));

            // Throttle per-AMS evaluation.
            var now = DateTime.UtcNow;
            if (_lastEvalAt.TryGetValue(snapshot.AmsId, out var last)
                && now - last < cfg.TickInterval)
                continue;
            _lastEvalAt[snapshot.AmsId] = now;

            // One controller per AMS, lazily created.
            if (!_controllers.TryGetValue(snapshot.AmsId, out var controller))
            {
                controller = new AutoDryController(snapshot.AmsId);
                _controllers[snapshot.AmsId] = controller;
            }

            var settings = cfg.SettingsByAmsId.TryGetValue(snapshot.AmsId, out var s) ? s : cfg.DefaultSettings;
            var decision = controller.Evaluate(snapshot, settings, isPrinting);
            Emit(new Event.DecisionMade(DateTime.UtcNow, snapshot.AmsId, decision));

            switch (decision)
            {
                case Decision.Start start:
                    if (cfg.DryRun)
                    {
                        Emit(new Event.DryRunSkipped(DateTime.UtcNow, start.AmsId, decision));
                    }
                    else
                    {
                        try
                        {
                            var json = await _transport.StartDryingAsync(
                                start.AmsId, start.Filament, start.TempC, start.Hours, ct).ConfigureAwait(false);
                            Emit(new Event.Sent(DateTime.UtcNow, start.AmsId, json));
                        }
                        catch (Exception ex)
                        {
                            Emit(new Event.ErrorOccurred(DateTime.UtcNow, $"start publish failed: {ex.Message}"));
                        }
                    }
                    break;

                case Decision.Stop stop:
                    if (cfg.DryRun)
                    {
                        Emit(new Event.DryRunSkipped(DateTime.UtcNow, stop.AmsId, decision));
                    }
                    else
                    {
                        try
                        {
                            var json = await _transport.StopDryingAsync(stop.AmsId, ct).ConfigureAwait(false);
                            Emit(new Event.Sent(DateTime.UtcNow, stop.AmsId, json));
                        }
                        catch (Exception ex)
                        {
                            Emit(new Event.ErrorOccurred(DateTime.UtcNow, $"stop publish failed: {ex.Message}"));
                        }
                    }
                    break;
            }
        }
    }

    private void Emit(Event e) => _events.Writer.TryWrite(e);

    private static IReadOnlyDictionary<int, AutoDrySettings> With(IReadOnlyDictionary<int, AutoDrySettings> src, int key, AutoDrySettings value)
    {
        var copy = new Dictionary<int, AutoDrySettings>(src) { [key] = value };
        return copy;
    }
}

public sealed record OrchestratorConfig
{
    public required IReadOnlyDictionary<int, AutoDrySettings> SettingsByAmsId { get; init; }
    public required AutoDrySettings DefaultSettings { get; init; }
    public bool DryRun { get; init; } = true;
    public TimeSpan TickInterval    { get; init; } = TimeSpan.FromSeconds(5);
    public TimeSpan PushAllInterval { get; init; } = TimeSpan.FromSeconds(60);

    public static OrchestratorConfig Default() => new()
    {
        SettingsByAmsId = new Dictionary<int, AutoDrySettings>(),
        DefaultSettings = new AutoDrySettings(),
    };
}
